bookwyrm/views/preferences/export.py

```diff
@@ -152,6 +154,36 @@ def get(self, request):
         jobs = BookwyrmExportJob.objects.filter(user=request.user).order_by(
             "-created_date"
         )
+
+        exports = []
+        for job in jobs:
+            export = {"job": job}
+
+            if settings.USE_S3:
+                # make custom_domain None so we can sign the url
+                # see https://github.com/jschneier/django-storages/issues/944
+                storage = S3Boto3Storage(querystring_auth=True, custom_domain=None)
+
+                # for s3 we download directly from s3, so we need a signed url
+                export["url"] = S3Boto3Storage.url(
+                    storage, f"/exports/{job.task_id}.tar.gz", expire=900
+                )  # temporarily downloadable file, expires after 5 minutes
```

Comment:
comment says 5 minutes, but 900 is 15 minutes in seconds. Does either figure need updating?

Reply:
hmm, honestly I'm not sure which I intended. 5 minutes seems reasonable?

Reply:
5 minutes seems tight to me, I can imagine getting up for a glass of water and missing it.

Reply:
@mouse-reeve @hughrun Especially if the process is asynchronous and we don't exactly know when it ends. For users with large archive it could be easy to miss even probably even in 15 minutes.

Reply:
Would it make sense to have a settings.py/env variable for this with a reasonable default?

Reply:
Also I've checked Mastodon's source. Didn't find the lifetime of the link there, but I've found out that they're also adding some randomness to the archive filename. Which seems reasonable in case of local storage for example.

[app/services/backup_service.rb:64](https://github.com/mastodon/mastodon/blob/81a04ac25cb5a25d8501e59a16b272a6b4e177f0/app/services/backup_service.rb#L64)

Reply:
Ok yes 5 mins is too short but I think 7 days is excessive. The url will use a random uuid but we have to remember this is still a file potentially containing private statuses.

I'll fix this up probably tomorrow and make it configurable.

Reply:
I'm just pointing at a fact that as this is asynchronous process, user may not even be able to download their archive the same day 🤷

I didn't see an entire code (sorry), but if users just gets new URL when trying to download their archive it's fine. But if they need to wait for new archive generation due to expired link, it may become an endless loop for them.

- request archive
- it takes too long
- I'll check out later
- link expired
- requesting new archive
- ...

Reply:
Ok folks I think there is some confusion here.

This is a completely separate function from the creation of a user export file. That process is triggered when the user requests the export. It's a series of background jobs: making it a synchronous process would be bonkers.

What we are looking at here is the function that runs _every time the user loads the user exports page_. That's why I originally was thinking 5 minutes would be reasonable: if it has expired you can reload the page and get a fresh link for every available download file you've generated.

Reply:
If you want to avoid ever letting the user face an expiring link, another option is to make the link in the HTML point to an internal (authenticated) endpoint that redirects to the signed S3 link. That way the link has to be valid only for a very short time, since it is created immediately before redirecting the user to it.

(But with @hughrun's explanation, generating a download link that is valid for a few minutes when the user opens the exports page is definitely fine too)

Reply:
@hughrun 
> What we are looking at here is the function that runs every time the user loads the user exports page

Ah. Thanks for clearing this out! Then my comments are mostly non-relevant 😞

> make the link in the HTML point to an internal (authenticated) endpoint that redirects to the signed S3 link

I double that. Requesting new links form S3 every time user loads the page seems to have at least two downsides:

- You can't send such link via email for example like Mastodon does
- It'll cause useless S3 requests which could be non-free in most of object storage providers

Reply:
I implemented the redirect for S3.